Mathlib/CategoryTheory/Bicategory/Monad/Basic.lean

```diff
@@ -0,0 +1,165 @@
+
+@[inherit_doc] scoped notation "ε" => Comonad.counit
+@[inherit_doc] scoped notation "ε["x"]" => Comonad.counit (t := x)
+@[inherit_doc] scoped notation "Δ" => Comonad.comul
+@[inherit_doc] scoped notation "Δ["x"]" => Comonad.comul (t := x)
+
+namespace Comonad
+
+variable {a: B}
+
+/- Comonad laws -/
+section
+
+variable (t : a ⟶ a) [Comonad t]
+
+@[reassoc (attr := simp)]
+theorem counit_comul : Δ ≫ ε ▷ t = (λ_ t).inv := Comon_Class.counit_comul t
+
+@[reassoc (attr := simp)]
+theorem comul_counit : Δ ≫ t ◁ ε = (ρ_ t).inv := Comon_Class.comul_counit t
+
+@[reassoc (attr := simp)]
+theorem comul_assoc : Δ ≫ t ◁ Δ = Δ ≫ Δ ▷ t ≫ (α_ t t t).hom := Comon_Class.comul_assoc t
+
+@[reassoc]
+theorem comul_assoc_flip : Δ ≫ Δ ▷ t = Δ ≫ t ◁ Δ ≫ (α_ t t t).inv := Comon_Class.comul_assoc_flip t
+
+end
+
+@[simps! counit]
+instance {a : B} : Comonad (𝟙 a) :=
+  inferInstanceAs <| Comon_Class (MonoidalCategory.tensorUnit (a ⟶ a))
+
+/-- An oplax functor from the trivial bicategory to `B` defines a comonad in `B`. -/
+def ofOplaxFromPUnit (F : OplaxFunctor (LocallyDiscrete (Discrete PUnit.{u₀ + 1})) B) :
+    Comonad (F.map (𝟙 ⟨⟨PUnit.unit⟩⟩)) where
+  comul := F.map₂ (ρ_ _).inv ≫ F.mapComp _ _
+  counit := F.mapId _
+  comul_assoc' := by
+    simp only [tensorObj_def, MonoidalCategory.whiskerLeft_comp, whiskerLeft_def, Category.assoc,
+      MonoidalCategory.comp_whiskerRight, whiskerRight_def, associator_def]
+    rw [← F.mapComp_naturality_left_assoc, ← F.mapComp_naturality_right_assoc]
+    simp only [whiskerLeft_rightUnitor_inv, PrelaxFunctor.map₂_comp, Category.assoc,
+      OplaxFunctor.map₂_associator, whiskerRight_id, Iso.hom_inv_id_assoc]
+  counit_comul' := by
+    simp only [tensorUnit_def, tensorObj_def, whiskerRight_def, Category.assoc, leftUnitor_def]
+    rw [F.mapComp_id_left, unitors_equal, F.map₂_inv_hom_assoc]
+  comul_counit' := by
+    simp only [tensorUnit_def, tensorObj_def, whiskerLeft_def, rightUnitor_def]
+    rw [Category.assoc, F.mapComp_id_right, F.map₂_inv_hom_assoc]
+
+/-- A comonad in `B` defines an oplax functor from the trivial bicategory to `B`. -/
+def toOplax {a : B} (t : a ⟶ a) [Comonad t] :
+    OplaxFunctor (LocallyDiscrete (Discrete PUnit.{u₀ + 1})) B where
+  obj _ := a
+  map _ := t
+  map₂ _ := 𝟙 _
+  mapId _ := ε
+  mapComp _ _ := Δ
+  map₂_associator _ _ _ := by
+    rw [Category.id_comp]
+    apply Comonad.comul_assoc
+
+end Comonad
+
+/- In this section, we define bicategory structure on comonads by using the bicategory structure on
+oplax functors. We may use oplax, lax, or pseudonatural transformations to provide the bicategory
+structure, and the namespace below indicates that we use oplax transformations here. The
+constructions for the other two cases would be given in the corresponding namespaces. -/
+namespace OplaxTrans
+
+/-- The bicategory of comonads in `B`. -/
+def ComonadBicat (B : Type u) [Bicategory.{w, v} B] :=
+  OplaxFunctor (LocallyDiscrete (Discrete PUnit.{u₀ + 1})) B
```

Comment:
Could you explain why you introduce an auxiliary universe `u₀` here?

Reply:
Actually, I don't have any particular preference. I'm following mathlib conventions, in which PUnit is used more widely than Unit (PUnit with `u₀ = 0`).

Reply:
We certainly use `PUnit` when we want a topological space in `Type u` for a *given* "target" universe `u`, etc. But otherwise, I do not see any reason to introduce an extraneous universe.